cmd/vic-machine/converter/converter.go

```diff
@@ -0,0 +1,506 @@
+		if src.Int() == 0 {
+			return nil
+		}
+		v = strconv.FormatInt(src.Int(), 10)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		if src.Uint() == 0 {
+			return nil
+		}
+		v = strconv.FormatUint(src.Uint(), 10)
+	case reflect.Bool:
+		if !src.Bool() {
+			return nil
+		}
+		v = strconv.FormatBool(src.Bool())
+	case reflect.Float32, reflect.Float64:
+		if src.Float() == 0 {
+			return nil
+		}
+		v = strconv.FormatFloat(src.Float(), 'E', -1, 64)
+	default:
+		panic(fmt.Sprintf("%s is not supported type", src.Kind()))
+	}
+	log.Debugf("%s=%s", prefix, v)
+
+	addValue(dest, prefix, v)
+	return nil
+}
+
+// convertNetwork will merge destination and gateway to one option with format: 192.168.3.0/16,192.168.128.0/16:192.168.2.1
+// after that, convertStruct is called for left conversion
+func convertNetwork(src reflect.Value, prefix string, tags reflect.StructTag, dest map[string][]string) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("prefix: %s, src: %s", prefix, src.String())))
+
+	if prefix == "" {
+		return nil
+	}
+	if tags.Get(cmdTag) == "" {
+		return nil
+	}
+
+	network, ok := src.Interface().(data.NetworkConfig)
+	if !ok {
+		panic(fmt.Sprintf(src.Type().String() + " is not NetworkConfig"))
+	}
+	if !network.IsSet() {
+		return nil
+	}
+
+	if len(network.Destinations) > 0 || !ip.Empty(network.Gateway) {
+		destination := ""
+		if len(network.Destinations) > 0 {
+			for _, d := range network.Destinations {
+				destination = fmt.Sprintf("%s,%s", destination, d.String())
+			}
+			destination = strings.TrimLeft(destination, ",")
+		}
+		gateway := network.Gateway.IP.String()
+		tag := "cmd:\"gateway\""
+		key := calculateKey(reflect.StructTag(tag), prefix)
+		if destination != "" {
+			addValue(dest, key, fmt.Sprintf("%s:%s", destination, gateway))
+		} else {
+			addValue(dest, key, gateway)
+		}
+	}
+	return convertStruct(reflect.ValueOf(network), prefix, tags, dest)
+}
+
+// convertContainerNetworks will switch the map keys in MappedNetworks using value, and replace all keys with the same value in other structure,
+// cause option is using vsphere network name as key label, but guestinfo is using alias as key for easy to use in portlayer
+// after that, convertStruct is called for left conversion
+func convertContainerNetworks(src reflect.Value, prefix string, tags reflect.StructTag, dest map[string][]string) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("prefix: %s, src: %s", prefix, src.String())))
+
+	if prefix == "" {
+		return nil
+	}
+	if tags.Get(cmdTag) == "" {
+		return nil
+	}
+
+	networks, ok := src.Interface().(data.ContainerNetworks)
+	if !ok {
+		panic(fmt.Sprintf(src.Type().String() + " is not ContainerNetworks"))
+	}
+	if !networks.IsSet() {
+		return nil
+	}
+
+	for k, v := range networks.MappedNetworks {
+		if k == v {
+			continue
+		}
+		if dns, ok := networks.MappedNetworksDNS[k]; ok {
+			networks.MappedNetworksDNS[v] = dns
+			delete(networks.MappedNetworksDNS, k)
```

Comment:
Why are we modifying the input here? Generally for code like this, the input is not modified. If any modifications are needed, then operate on a copy.